### ISSUE TITLE
Actions in cluster mode

### ISSUE THREAD

**Issue by nicokruger** [keymetrics/pmx#75]
Are actions in cluster mode supposed to execute over all workers in the cluster?

I've managed to create a simple action, however it only seems to be running on the first worker in the cluster (cluster managed by pm2).

**Comment by vmarchaud** [comment]
No, actions are supposed to run on only one worker (and indeed only the first) but we can code something to trigger action on all worker if we want. Can you make a issue over PM2 repository to make an options for this ?

**Comment by nicokruger** [comment]
Great stuff, thanks for the really quick reply.

I was coming from the use case of keymetrics.io - I have now since figured out that if you filter on keymetrics.io by the server and the app, and then click the "top" action button it executes it on all workers. If you don't find filter, but just select "Show" on your app and then click the action button there, it only does it on the first.

Not sure if I'm making sense and it's a bit unintuitive but not sure what the site is doing. I'll play with it a bit more and then see if its needed for a feature request on the PM2 side?

Also not sure if it's maybe worth getting the keymetrics guys to chip in as to what exactly they are doing differently depending on which button you press.

**Comment by vmarchaud** [comment]
Yeah totally, i was thinking that you were coming from PM2 (so you were using the CLI command `pm2 trigger`), the problem here is that the dashboard (without filter) only show you the first worker and then under all the workers (if you look, the big one is the same as the first in the details) so obviously if you run the command it will only run on the worker that is shown by the dashboard.

The way we designed the dashboard is that you mainly want to only make an action to one worker (like getting some stats from any database) so currently the dashboard isn't designed for making the action executing by default on all workers (you can't choose to focus on only one worker).

So at least for now, i guess it would stay like this (but i will discuss with our team tomorrow about this)

PS : Actually i'm a keymetrics guys

**Comment by nicokruger** [comment]
Ok cool. BTW really digging keymetrics :)

I get that you'd want the ability to only execute the action on one worker, for the stats example. Our use case is more like the "set a debug flag" use case - it is an action that needs to be run on all.

And then just to throw another spanner in the works - we also have another use case where we just want worker to reply (similair to getting some stats from the database).

Basically I've figured out how to do what we want for now, it's just you need to wangle the UI state first in different ways for the first and second use case, and I'm worried that some of our ops guys will not get that right all the time.

Anyway, thanks for all the great on keymetrics!

**Comment by vmarchaud** [comment]
Yeah totally i get your point, we have planned to work on the dashboard, it will be on the list of things to think of.
The flags isn't actually needed since its the dashboard that ask for the action on only one process cluster, we just need to find a solution to show somewhere an option to run the action on all workers.
If you have more questions about our product, i suggest you to come directly on our slack (http://slack.pm2.io) and directly ask me in private (same username as github).

**Comment by nicokruger** [comment]
Great stuff, thanks.

This can be closed theb.
